filebeat/input/tcp/client.go

```diff
@@ -0,0 +1,116 @@
+package tcp
+
+import (
+	"bufio"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/filebeat/harvester"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Client is a remote client.
+type Client struct {
+	conn           net.Conn
+	forwarder      *harvester.Forwarder
+	done           chan struct{}
+	metadata       common.MapStr
+	splitFunc      bufio.SplitFunc
+	maxReadMessage uint64
+	timeout        time.Duration
+}
+
+// NewClient returns a new client instance for the remote connection.
+func NewClient(
+	conn net.Conn,
+	forwarder *harvester.Forwarder,
+	splitFunc bufio.SplitFunc,
+	maxReadMessage uint64,
+	timeout time.Duration,
+) *Client {
+	client := &Client{
+		conn:           conn,
+		forwarder:      forwarder,
+		done:           make(chan struct{}),
+		splitFunc:      splitFunc,
+		maxReadMessage: maxReadMessage,
+		timeout:        timeout,
+	}
+	client.cacheMetadata()
+	return client
+}
+
+// Handle is reading message from the specified TCP socket.
+func (c *Client) Handle() error {
+	r := NewResetableLimitedReader(NewDeadlineReader(c.conn, c.timeout), c.maxReadMessage)
+	buf := bufio.NewReader(r)
+	scanner := bufio.NewScanner(buf)
+	scanner.Split(c.splitFunc)
+
+	for scanner.Scan() {
+		err := scanner.Err()
+		if err != nil {
+			// we are forcing a close on the socket, lets ignore any error that could happen.
+			select {
+			case <-c.done:
+				break
+			default:
+			}
+			// This is a user defined limit and we should notify the user.
+			if IsMaxReadBufferErr(err) {
+				logp.Err("tcp client error: %s", err)
+			}
+			return errors.Wrap(err, "tcp client error")
+		}
+		r.Reset()
+		c.forwarder.Send(c.createEvent(scanner.Text()))
+	}
+	return nil
+}
+
+// Close stops reading from the socket and close the connection.
+func (c *Client) Close() {
+	close(c.done)
+	c.conn.Close()
+}
+
+func (c *Client) createEvent(rawString string) *util.Data {
+	data := util.NewData()
+	data.Event = beat.Event{
+		Timestamp: time.Now(),
+		Meta:      c.metadata,
```

Comment:
The `metadata` field behaves like `@metadata` in Logstash. If one sets `index` or `id` or `pipeline`, this will be configure some paramters in the Elasticsearch output. This on purpose?

Reply:
I did it on purpose, I am not sure how valuable that information it is. but I could make it public field and store in the mapping. WDYT?